### PR TITLE
Proposal <-> Thread Linking 

### PR DESCRIPTION
**Pull request by ianrowan**
## Link to Issue
Closes: #3357 

#### Backend Changes
- Database migration to add JSONB links column to Threads + consolidate existing links(snapshot, threads, chainEntityMeta) into new column for existing records
- New routes
    - `POST /linking/getLinks`: get bi-directional links for threads <-> (threads, snapshot, proposals)
    - `POST /linking/addThreadLinks`: add Links to a existing thread_id
    - `DELETE /linking/deleteLinks`: delete links from Thread Id
- API integration tests 
- Add titles to links and aggregate existing titles
- Update `bulkThreads` route to remove chain events and other deprecated link dependencies

#### Frontend Changes
- deprecated `thread.snapshotProposal` and `thread.linkedThreads` in favour of `thread.links`
- updated flow within multiple components that are responsible for handling linking threads/snapshots and proposals
- used new methods in threads controller (`addLinks`, `deleteLinks`, `getThreadsForLink`) and removed several 
- stubs for legacy link building
- Deprecate `snapshot_proposal` on thread
- Deprecate `linked_threads` on thread

#### Backend
- Full coverage Integrations tests(added) must pass. 
- Frontend must not experience issues with new routes 

#### Frontend
- create thread and try to add proposals, snapshots and threads
- click around links on a thread 
- ensure old links are on proposals(prior to any created on test deployment) 

## Deployment Plan

1. Run migration

## Other Considerations

- Issues related to missing old snapshot records are still outstanding and may cause issues on those older threads if links are still in use #3611
- Older chainEntityMeta Ids stored on links will need to be retroactively updated in some way to include hyperlinks similar to how all new proposals(cosmos, etc) will be store for quick linking: #3639 
- Old tables/columns will still exist but will stop receiving data that is being stored in `links`. We should decide to archive or deprecate these 
- Potential addition of update methods after consideration of how access controls will work here and how it will be used
- Additional implementation req. for use with cosmos/non-ce/snapshot

### CODE REVIEW

File: packages/commonwealth/client/scripts/controllers/server/threads.ts [hunk 1]
```diff
@@ -516,7 +517,90 @@ class ThreadsController {
...
+  /**
+   * Add a set of links to a Thread
+   * @param args 
+   * @returns updated Thread
+   */
+  public async addLinks (args: {
+    threadId: number;
+    links: link[];
+  }): Promise<Thread>{
+    const response = await $.ajax({
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/controllers/server/threads.ts:529]: please do not use jquery for API calls, use axios instead (everywhere)

Comment (ianrowan) [reply packages/commonwealth/client/scripts/controllers/server/threads.ts:529]: @masvelio Good catch. Just copied the existing calls in this file with new links, but I'll convert the new ones to axios

File: packages/commonwealth/client/scripts/models/Thread.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import { ProposalType } from 'common-common/src/types';
 import type moment from 'moment';
+import { link } from 'server/models/thread';
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/models/Thread.ts:3]: Please do not import files from the server on the client, as we don't want to mix client and server code / imports. Ideally we should recreate the type on the client (or maybe import type will work fine, TBD).

File: packages/commonwealth/client/scripts/controllers/server/threads.ts [hunk 3]
```diff
@@ -24,6 +24,7 @@ import app from 'state';
 import { ProposalStore, RecentListingStore } from 'stores';
 import { orderDiscussionsbyLastComment } from 'views/pages/discussions/helpers';
 import { EventEmitter } from 'events';
+import { link, linkSource } from 'server/models/thread';
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/server/threads.ts:27]: See comment below about importing from server.

File: packages/commonwealth/server/models/thread.ts [hunk 4]
```diff
@@ -95,6 +108,7 @@ export default (
         allowNull: false,
       },
       snapshot_proposal: { type: dataTypes.STRING(48), allowNull: true },
+      inks: {type: dataTypes.JSONB, allowNull: true},
```
Comment (masvelio) [inline packages/commonwealth/server/models/thread.ts:111]: `inks` => `links`
Also, I think we need some migration to create minimal dataset (or even empty tables with column). Right now, the app on the branch is not usable
image

File: packages/commonwealth/server/models/thread.ts [hunk 5]
```diff
@@ -8,6 +8,18 @@ import type { LinkedThreadAttributes } from './linked_thread';
 import type { TopicAttributes } from './topic';
 import type { ModelInstance, ModelStatic } from './types';
 
+export enum linkSource {
```
Comment (masvelio) [inline packages/commonwealth/server/models/thread.ts:11]: `linkSource` => `LinkSource`
look like this is our convention

File: packages/commonwealth/server/models/thread.ts [hunk 6]
```diff
@@ -8,6 +8,18 @@ import type { LinkedThreadAttributes } from './linked_thread';
...
+  Snapshot = 'snapshot',
+  Proposal = 'proposal',
+  Thread = 'thread',
+  Web = 'web'
+}
+
+export type link = {
```
Comment (masvelio) [inline packages/commonwealth/server/models/thread.ts:18]: same here `link` => `Link`

File: packages/commonwealth/test/unit/proposal_linking/proposalLinking.spec.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        assert.equal(response[0], 'proposal/1')
+    })
+
+    it('should get ThreadsForLink', async () => {
+        const threadInst = thread.getById(2)
+        const threadsLinked = await thread.getThreadsForLink({link: {source: linkSource.Proposal, identifier: '1'}})
+        assert.equal(threadsLinked[0], threadInst)
+    })
+    
+  });
```
Comment (masvelio) [inline packages/commonwealth/test/unit/proposal_linking/proposalLinking.spec.ts:37]: make sure you have a prettier configured on your IDE properly
see this comment https://github.com/hicommonwealth/commonwealth/pull/3423#issuecomment-1506648147

File: packages/commonwealth/client/scripts/controllers/server/threads.ts [hunk 8]
```diff
@@ -516,7 +517,90 @@ class ThreadsController {
...
+   * @param args 
+   * @returns updated Thread
+   */
+  public async deleteLinks (args: {
+    threadId: number;
+    links: link[];
+  }): Promise<Thread> {
+    const response = await $.ajax({
+      url: `${app.serverUrl()}/linking/deleteLinks`,
+      type: 'POST',
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/controllers/server/threads.ts:552]: probably DELETE here would be more suitable

File: packages/commonwealth/server/models/thread.ts (hunk 4 above)
Comment (ianrowan) [reply packages/commonwealth/server/models/thread.ts]: Added the migration to add the column here from BE branch + fixed links

File: packages/commonwealth/server/models/thread.ts (hunk 5 above)
Comment (ianrowan) [reply packages/commonwealth/server/models/thread.ts]: fixed this

File: packages/commonwealth/server/models/thread.ts (hunk 6 above)
Comment (ianrowan) [reply packages/commonwealth/server/models/thread.ts]: fixed this

File: packages/commonwealth/test/unit/proposal_linking/proposalLinking.spec.ts (hunk 7 above)
Comment (ianrowan) [reply packages/commonwealth/test/unit/proposal_linking/proposalLinking.spec.ts:37]: formatted

**Comment by jnaviask** [comment]
Amazing. @ianrowan can we push this to frick or frack for QA?

**Comment by ianrowan** [comment]
@jnaviask pushed to frick for QA

**Comment by jnaviask** [comment]
cc @ianrowan @masvelio: Tested on QA. Everything works on dYdX.

Main issue I uncovered is that on e.g. Juno, the "Link Proposal" button pops up the "update thread stage" modal, which does not allow me to link a Juno on-chain proposal as expected. We need to make sure this is fixed, as supporting easy Cosmos-base proposal-to-thread linking is one of the top wins from this PR.

**Review by jnaviask — APPROVED** [review]
LETS GO!!! Waiting on this bugfix followup #3683 for "ready to roll", but we can get this out the door now. Great work yall!

**Pull request merged**